<a href="https://colab.research.google.com/github/fomen2001/j-apprends-les-bases-de-la-data-science/blob/main/Valdes_Joel_FOMENA_TSATSOP_Analyse_Bitcoin.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Chargement des donnees**

In [407]:
# Import des librairies pandas et numpy.
import pandas as pd
import numpy as np

In [408]:
# Lecture des données stockés au format csv

btc = pd.read_csv(
    "BTC_Historical_Data.csv", sep=';', decimal=','
)

In [409]:
btc

Date         Open         High          Low        Close  \
0     2008-07-09 00:00:00      0.00000      0.00000      0.00000      0.00000   
1     2008-07-10 00:00:00      0.00000      0.00000      0.00000      0.00000   
2     2008-07-11 00:00:00      0.00000      0.00000      0.00000      0.00000   
3     2008-07-12 00:00:00      0.00000      0.00000      0.00000      0.00000   
4     2008-07-13 00:00:00      0.00000      0.00000      0.00000      0.00000   
...                   ...          ...          ...          ...          ...   
5826  2025-03-08 00:00:00  86154.59375  86847.26563  85247.48438  86742.65625   
5827  2025-03-09 00:00:00  80601.03906  86471.13281  80052.48438  86154.30469   
5828  2025-03-10 00:00:00  78532.00000  83955.92969  77420.59375  80597.14844   
5829  2025-03-11 00:00:00  82862.21094  83577.75781  76624.25000  78523.87500   
5830  2025-03-12 00:00:00  83722.35938  84358.57813  80635.25000  82857.37500   

            Volume  
0     0.000000e+00  
1     0.000000e+00  
2     0.000000e+00  
3     0.000000e+00  
4     0.000000e+00  
...            ...  
5826  1.820612e+10  
5827  3.089935e+10  
5828  5.406110e+10  
5829  5.470284e+10  
5830  4.035348e+10  

[5831 rows x 6 columns]

In [435]:
btc.shape

(5093, 5)

In [410]:
#J'effectue un filtre pour la variable close pour eleminer les donnees de la variable egal a 0
btc = btc[(btc[['Close']] != 0).all(axis=1)]


In [411]:
btc.head(15)

Date     Open     High      Low    Close    Volume
738  2010-07-17 00:00:00  0.04951  0.04951  0.04951  0.04951    0.9902
739  2010-07-18 00:00:00  0.04951  0.08585  0.04951  0.08584    5.0920
740  2010-07-19 00:00:00  0.08584  0.09307  0.07723  0.08080   49.6600
741  2010-07-20 00:00:00  0.08080  0.08181  0.07426  0.07474   20.5900
742  2010-07-21 00:00:00  0.07474  0.07921  0.06634  0.07921   42.2600
743  2010-07-22 00:00:00  0.07921  0.08181  0.05050  0.05842  129.7800
744  2010-07-23 00:00:00  0.05842  0.06767  0.05050  0.06262  141.0700
745  2010-07-24 00:00:00  0.06262  0.06262  0.05049  0.05454   26.7300
746  2010-07-25 00:00:00  0.05454  0.05941  0.05050  0.05050   85.0600
747  2010-07-26 00:00:00  0.05050  0.05600  0.05000  0.05600   46.9100
748  2010-07-27 00:00:00  0.05600  0.06050  0.05300  0.06000  196.9200
749  2010-07-28 00:00:00  0.06000  0.06200  0.05400  0.05890  255.7600
750  2010-07-29 00:00:00  0.05890  0.06990  0.05710  0.06990  528.3200
751  2010-07-30 00:00:00  0.06990  0.06990  0.05820  0.06270  198.5300
752  2010-07-31 00:00:00  0.06270  0.06889  0.05600  0.06785  243.9000

In [412]:
btc.head()

Date     Open     High      Low    Close   Volume
738  2010-07-17 00:00:00  0.04951  0.04951  0.04951  0.04951   0.9902
739  2010-07-18 00:00:00  0.04951  0.08585  0.04951  0.08584   5.0920
740  2010-07-19 00:00:00  0.08584  0.09307  0.07723  0.08080  49.6600
741  2010-07-20 00:00:00  0.08080  0.08181  0.07426  0.07474  20.5900
742  2010-07-21 00:00:00  0.07474  0.07921  0.06634  0.07921  42.2600

In [413]:
btc.dtypes

Date       object
Open      float64
High      float64
Low       float64
Close     float64
Volume    float64
dtype: object

In [414]:
# Formatage du champ date
btc["Date"] = pd.to_datetime(
      btc["Date"]
    #, format="%Y-%m-%d", errors="coerce"
)
# Extraire uniquement la date sans l'heure
#btc['Date'] = btc['Date'].dt.date


# Afficher les premières lignes du DataFrame pour vérifier


btc

<ipython-input-414-3b74141a20bb>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  btc["Date"] = pd.to_datetime(


Date         Open         High          Low        Close  \
738  2010-07-17      0.04951      0.04951      0.04951      0.04951   
739  2010-07-18      0.04951      0.08585      0.04951      0.08584   
740  2010-07-19      0.08584      0.09307      0.07723      0.08080   
741  2010-07-20      0.08080      0.08181      0.07426      0.07474   
742  2010-07-21      0.07474      0.07921      0.06634      0.07921   
...         ...          ...          ...          ...          ...   
5826 2025-03-08  86154.59375  86847.26563  85247.48438  86742.65625   
5827 2025-03-09  80601.03906  86471.13281  80052.48438  86154.30469   
5828 2025-03-10  78532.00000  83955.92969  77420.59375  80597.14844   
5829 2025-03-11  82862.21094  83577.75781  76624.25000  78523.87500   
5830 2025-03-12  83722.35938  84358.57813  80635.25000  82857.37500   

            Volume  
738   9.902000e-01  
739   5.092000e+00  
740   4.966000e+01  
741   2.059000e+01  
742   4.226000e+01  
...            ...  
5826  1.820612e+10  
5827  3.089935e+10  
5828  5.406110e+10  
5829  5.470284e+10  
5830  4.035348e+10  

[5093 rows x 6 columns]

In [415]:
btc.dtypes

Date      datetime64[ns]
Open             float64
High             float64
Low              float64
Close            float64
Volume           float64
dtype: object

In [416]:
#verifions les objets que nous avons
btc.keys()

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Volume'], dtype='object')

In [417]:
# Voir les indexes de notre dataset
btc.index

Index([ 738,  739,  740,  741,  742,  743,  744,  745,  746,  747,
       ...
       5821, 5822, 5823, 5824, 5825, 5826, 5827, 5828, 5829, 5830],
      dtype='int64', length=5093)

In [418]:
#definir la date comme index
btc.set_index('Date', inplace=True)

In [419]:
btc.head()

Open     High      Low    Close   Volume
Date                                                   
2010-07-17  0.04951  0.04951  0.04951  0.04951   0.9902
2010-07-18  0.04951  0.08585  0.04951  0.08584   5.0920
2010-07-19  0.08584  0.09307  0.07723  0.08080  49.6600
2010-07-20  0.08080  0.08181  0.07426  0.07474  20.5900
2010-07-21  0.07474  0.07921  0.06634  0.07921  42.2600

In [436]:
#affichons les colonnes close et volume
btc_subset = btc[['Close', 'Volume']]
btc_subset.head(15)


Close    Volume
Date                         
2010-07-17  0.04951    0.9902
2010-07-18  0.08584    5.0920
2010-07-19  0.08080   49.6600
2010-07-20  0.07474   20.5900
2010-07-21  0.07921   42.2600
2010-07-22  0.05842  129.7800
2010-07-23  0.06262  141.0700
2010-07-24  0.05454   26.7300
2010-07-25  0.05050   85.0600
2010-07-26  0.05600   46.9100
2010-07-27  0.06000  196.9200
2010-07-28  0.05890  255.7600
2010-07-29  0.06990  528.3200
2010-07-30  0.06270  198.5300
2010-07-31  0.06785  243.9000

**correlation close et volume**

In [437]:
# Calcul de la corrélation entre "Close" et "Volume"

btc_monthly = btc_subset.resample('ME').mean()
correlation = btc_subset['Close'].corr(btc['Volume'])

print(f"Corrélation entre Close et Volume : {correlation}")

Corrélation entre Close et Volume : 0.7261087606413652


La corrélation entre les variables "Close" et "Volume" est de 0.726, ce qui indique une relation positive modérément forte entre ces deux variables

In [438]:
#afficher toutes les correlations
btc_subset.corr()

Close    Volume
Close   1.000000  0.726109
Volume  0.726109  1.000000

In [423]:
#correlation entre close et volume
btc[['Close', 'Volume']].corr()

Close    Volume
Close   1.000000  0.726109
Volume  0.726109  1.000000

**Representation graphique de close et volume avec bokeh**


In [424]:
#importons la bibliotheque bokeh
from bokeh.plotting import figure, show
from bokeh.models import ColumnDataSource
from bokeh.io import output_notebook

# Affichage des graphiques directement dans le notebook
output_notebook()

In [482]:
from bokeh.models import ColumnDataSource, RangeTool
from bokeh.plotting import figure, show
from bokeh.layouts import column
import pandas as pd

# Vérifier que l'index est bien de type datetime
if not isinstance(btc.index, pd.DatetimeIndex):
    btc.index = pd.to_datetime(btc.index)

# Préparer la source de données
source = ColumnDataSource(data=dict(date=btc.index, close=btc['Close'], volume=btc['Volume']))

# Création du graphique pour Volume
p2 = figure(height=300, width=800, tools="xpan,wheel_zoom,box_zoom,reset",
            toolbar_location="above", x_axis_type="datetime",
            x_axis_location="above", background_fill_color="#efefef",
            x_range=(btc.index[0], btc.index[-1]))

p2.line('date', 'volume', source=source, line_width=2, color="green", legend_label="Volume")
p2.yaxis.axis_label = 'Volume'

# Ajout de la boîte de sélection
select = figure(title="Sélectionner la plage", height=130, width=800,
                y_range=p2.y_range, x_axis_type="datetime",
                y_axis_type=None, tools="", toolbar_location=None,
                background_fill_color="#efefef")

range_tool = RangeTool(x_range=p2.x_range)
range_tool.overlay.fill_color = "navy"
range_tool.overlay.fill_alpha = 0.2

select.line('date', 'volume', source=source, line_width=2, color="green")
select.ygrid.grid_line_color = None
select.add_tools(range_tool)

# Affichage des graphiques en colonne
show(column(p2, select))


In [481]:
from bokeh.models import ColumnDataSource, RangeTool
from bokeh.plotting import figure, show
from bokeh.layouts import column
import pandas as pd

# Vérifier que la colonne 'Date' est bien l'index et de type datetime
if not isinstance(btc.index, pd.DatetimeIndex):
    btc.index = pd.to_datetime(btc.index)

# Préparer la source de données
source = ColumnDataSource(data=dict(date=btc.index, close=btc['Close'], volume=btc['Volume']))

# Création du premier graphique pour Close avec zoom
p1 = figure(height=300, width=800, tools="xpan,wheel_zoom,box_zoom,reset",
            toolbar_location="above", x_axis_type="datetime",
            x_axis_location="above", background_fill_color="#efefef",
            x_range=(btc.index[0], btc.index[-1]))

p1.line('date', 'close', source=source, line_width=2, color="blue", legend_label="Close")
p1.yaxis.axis_label = 'Close'

# Ajout de la boîte de sélection
select = figure(title="Sélectionner la plage", height=130, width=800,
                y_range=p1.y_range, x_axis_type="datetime",
                y_axis_type=None, tools="", toolbar_location=None,
                background_fill_color="#efefef")

range_tool = RangeTool(x_range=p1.x_range)
range_tool.overlay.fill_color = "navy"
range_tool.overlay.fill_alpha = 0.2

select.line('date', 'close', source=source, line_width=2, color="blue")
select.ygrid.grid_line_color = None
select.add_tools(range_tool)

# Affichage des graphiques en colonne
show(column(p1, select))


In [477]:
from bokeh.models import HoverTool
# Créer la source de données pour Bokeh
source = ColumnDataSource(btc)

# Créer la figure
p = figure(title="Corrélation entre Close et Volume au fil du temps",
           x_axis_label="Date",
           y_axis_label="Volume",
           width=800, height=500,
           x_axis_type='datetime')  # Définir l'axe X comme étant de type datetime

# Ajouter le scatter plot avec la Date sur l'axe x et Volume sur l'axe y
p.scatter(x='Date', y='Volume', source=source, size=7, color="navy", alpha=0.6)

# Ajouter les tooltips pour afficher la Date, Close et Volume
tooltips = [
    ("Date", "@Date{%F}"),  # Afficher la date au format yyyy-mm-dd
    ("Close", "@Close{0.2f}"),
    ("Open", "@Open{0.2f}"),
    ("Volume", "@Volume{0.2f}")
]

# Ajouter le HoverTool pour afficher les tooltips
hover = HoverTool(tooltips=tooltips, formatters={'@Date': 'datetime'})
p.add_tools(hover)

# Affichage
output_notebook()  # Afficher le graphique directement dans un notebook Jupyter
show(p)

**Creation d'une serie mensuelle et annuelle sur la moyenne**

In [466]:
#serie mensuelle
btc_monthly = btc_subset.resample('ME').mean().interpolate(method='linear')

#serie annuelle
btc_annual = btc_subset.resample('YE').mean().interpolate(method='linear')

btc_monthly.head()
btc_annual.head()

Close        Volume
Date                                
2010-12-31    0.140399  2.581328e+03
2011-12-31    5.643498  2.356406e+05
2012-12-31    8.290571  4.170633e+05
2013-12-31  198.493434  7.747949e+06
2014-12-31  365.058217  2.383690e+07

In [464]:
import numpy as np
from bokeh.plotting import figure, show
from bokeh.models import ColumnDataSource, RangeTool
from bokeh.layouts import column

# Création de la série temporelle pour 'Close' et 'Volume'
dates = np.array(btc_monthly.index, dtype=np.datetime64)
source_close = ColumnDataSource(data=dict(date=dates, variable=btc_monthly['Close']))
source_volume = ColumnDataSource(data=dict(date=dates, variable=btc_monthly['Volume']))

# Graphique pour 'Close'
p_close = figure(height=300, width=800, tools="xpan", toolbar_location=None,
                 x_axis_type="datetime", x_axis_location="above",
                 background_fill_color="#efefef", x_range=(btc_subset.index[1500], btc_subset.index[2500]))

p_close.line('date', 'variable', source=source_close, color="blue")
p_close.yaxis.axis_label = 'Close'


# Création de la boîte de sélection
select = figure(title="Faites glisser le milieu et les bords de la boîte de sélection pour modifier la plage ci-dessus.",
                height=130, width=800, y_range=p_close.y_range,
                x_axis_type="datetime", y_axis_type=None,
                tools="", toolbar_location=None, background_fill_color="#efefef")

range_tool = RangeTool(x_range=p_close.x_range)
range_tool.overlay.fill_color = "navy"
range_tool.overlay.fill_alpha = 0.2

select.line('date', 'variable', source=source_close, color="blue")
select.ygrid.grid_line_color = None
select.add_tools(range_tool)


# Affichage du graphique
show(column(p_close, select))


In [461]:

import numpy as np
from bokeh.plotting import figure, show
from bokeh.models import ColumnDataSource, RangeTool
from bokeh.layouts import column

# Création de la série temporelle pour 'Close' et 'Volume'
dates = np.array(btc_monthly.index, dtype=np.datetime64)
source_close = ColumnDataSource(data=dict(date=dates, variable=btc_monthly['Close']))
source_volume = ColumnDataSource(data=dict(date=dates, variable=btc_monthly['Volume']))

# Graphique pour 'Volume'
p_volume = figure(height=300, width=800, tools="xpan", toolbar_location=None,
                  x_axis_type="datetime", x_axis_location="above",
                  background_fill_color="#efefef", x_range=(btc_subset.index[1500], btc_subset.index[2500]))

p_volume.line('date', 'variable', source=source_volume, color="green")
p_volume.yaxis.axis_label = 'Volume'
# Ajouter les tooltips pour afficher la Date, Close et Volume

# Création de la boîte de sélection
select = figure(title="Faites glisser le milieu et les bords de la boîte de sélection pour modifier la plage ci-dessus.",
                height=130, width=800, y_range=p_volume.y_range,
                x_axis_type="datetime", y_axis_type=None,
                tools="", toolbar_location=None, background_fill_color="#efefef")

range_tool = RangeTool(x_range=p_volume.x_range)
range_tool.overlay.fill_color = "navy"
range_tool.overlay.fill_alpha = 0.2

select.line('date', 'variable', source=source_close, color="green")
select.ygrid.grid_line_color = None
select.add_tools(range_tool)


# Affichage du graphique
show(column(p_volume, select))